File: cesta/SergioDeIscar-cesta-java/SergioDeIscar-cesta-java/src/main/java/org/example/models/Factura.java
package org.example.models;

import java.time.LocalDate;

public class Factura {
    final int id = count++;
    private static int count = 0;
    private final ItemFactura[] items = new ItemFactura[20];
    private String usuario;
    private LocalDate dateCreated;

    public Factura(String usuario) {
        if (usuario == null || usuario.isEmpty())
            throw new IllegalArgumentException("El usuario no puede ser nulo o vacío");
        setUsuario(usuario);
        dateCreated = LocalDate.now();
    }

    //region Getters
    public int getId() {
        return id;
    }
    public String getUsuario() {
        return usuario;
    }
    public LocalDate getDateCreated() {
        return dateCreated;
    }
    //endregion
    //region Setters
    public void setUsuario(String usuario) {
        if (usuario == null || usuario.isEmpty())
            throw new IllegalArgumentException("El usuario no puede ser nulo o vacío");
        this.usuario = usuario;
    }
    //endregion
    //region CRUD
    //region Add
    public ItemFactura addItem(ItemFactura newItem){
        int index = findIndexByItem(newItem);
        if (index != -1){
            items[index].setCantidad(items[index].getCantidad() + 1);
            return items[index];
        }

        for (int i = 0; i < items.length; i++) {
            if (items[i] == null){
                items[i] = newItem;
                return items[i];
            }
        }
        return null;
    }
    //endregion
    //region Remove
    public ItemFactura removeItem(int Index){
        if (Index < 0 || Index >= items.length)
            throw new IllegalArgumentException("El índice no puede ser menor a 0 o mayor a " + (items.length - 1));
        if (items[Index] == null) return null;
        ItemFactura item = items[Index];
        items[Index] = null;
        return item;
    }

    public ItemFactura removeItem(ItemFactura item){
        int index = findIndexByItem(item);
        if (index == -1) return null;
        return removeItem(index);
    }
    //endregion
    //region Update
    public ItemFactura updateItem(int Index, ItemFactura newItem){
        if (Index < 0 || Index >= items.length)
            throw new IllegalArgumentException("El índice no puede ser menor a 0 o mayor a " + (items.length - 1));
        if (items[Index] == null) return null;
        ItemFactura item = items[Index];
        items[Index] = newItem;
        return item;
    }

    public ItemFactura updateItem(ItemFactura item, ItemFactura newItem){
        int index = findIndexByItem(item);
        if (index == -1) return null;
        return updateItem(index, newItem);
    }
    //endregion
    //region Find
    public int findIndexByItem(ItemFactura item){
        for (int i = 0; i < items.length; i++) {
            if (items[i] == item){
                return i;
            }
        }
        return -1;
    }
    public ItemFactura findItem(int Index){
        if (Index < 0 || Index >= items.length)
            throw new IllegalArgumentException("El índice no puede ser menor a 0 o mayor a " + (items.length - 1));
        return items[Index];
    }
    public ItemFactura findItem(String name){
        for (ItemFactura item : items) {
            if (item != null && item.getProducto().getNombre().equals(name)){
                return item;
            }
        }
        return null;
    }
    //endregion
    //endregion
}

File: cesta/SergioDeIscar-cesta-kotlin/SergioDeIscar-cesta-kotlin/src/main/kotlin/models/Factura.kt
package factura.models

import java.time.LocalDate

class Factura(private val usuario: String, private val dateCreated: LocalDate) {
    val id = count++
    private val items: Array<ItemFactura?> = Array(20) { null }
    companion object {
        private var count = 0
    }

    constructor(usuario: String, dateCreate: LocalDate, items: Array<ItemFactura>): this(usuario, dateCreate){
        require(items.size in 1..this.items.size)
        for (i in items){
            addItem(i)
        }
    }

    //region CRUD
    // Add
    fun addItem(newItem: ItemFactura): ItemFactura?{
        val index = findIndexByItem(newItem)
        if (index != -1){
            items[index]?.cantidad = items[index]?.cantidad!! + 1
            return items[index]
        }

        for (i in items.indices){
            if (items[i] == null){
                items[i] = newItem
                return newItem
            }
        }
        return null
    }

    // Find
    fun findItem(name: String): ItemFactura?{
        for (i in items.indices){
            if (items[i] != null && items[i]?.item?.nombre == name){
                return items[i]
            }
        }
        return null
    }

    fun findItem(index: Int): ItemFactura?{
        require(index in items.indices)
        return items[index]
    }

    fun findItem(item: ItemFactura): ItemFactura?{
        return findItem(item.item.nombre)
    }

    fun findIndexByItem(item: ItemFactura): Int{
         for (i in items.indices){
             if (items[i] == item){
                 return i
             }
         }
        return -1
    }

    // Update
    fun updateItem(index: Int, newItem: ItemFactura): ItemFactura?{
        require(index in items.indices)
        if(items[index] != null) items[index] = newItem
        return items[index]
    }

    fun updateItem(oldItem: ItemFactura, newItem: ItemFactura): ItemFactura?{
        return updateItem(findIndexByItem(oldItem) ?: -1, newItem)
    }

    // Delete
    fun deleteItem(index: Int): Boolean{
        if (items[index] == null) return false
        items[index] = null
        return true
    }

    fun deleteItem(item: ItemFactura): Boolean{
        return deleteItem(findIndexByItem(item) ?: return false)
    }
    //endregion

    override fun toString(): String {
        val stringBuilder = StringBuilder("Factura ($id): Usuario -> $usuario, Date -> $dateCreated, Items -> ")
        for (i in items.indices){
            if (items[i] != null)
                stringBuilder.append("\n\t${i+1}->\t").append(items[i])
        }
        stringBuilder.append("\n\nValor Total:\t${totalValue()}")
        return stringBuilder.toString()
    }

    private fun totalValue(): Float {
        var count = 0.0f
        for (i in items){
            if (i != null)
                count += i.precioUnitario * i.cantidad
        }
        return count
    }
}